pkg/omitter/kubernetesresource.go

```diff
@@ -0,0 +1,114 @@
+package omitter
+
+import (
+	"encoding/json"
+	"errors"
+	"io/ioutil"
+	"strings"
+
+	"gopkg.in/yaml.v2"
+)
+
+type kubeMetadata struct {
+	Namespace string `yaml:"namespace" json:"namespace"`
+}
+
+type kubeResource struct {
+	ApiVersion string       `yaml:"apiVersion" json:"apiVersion"`
+	Kind       string       `yaml:"kind" json:"kind"`
+	Metadata   kubeMetadata `yaml:"metadata" json:"metadata"`
+}
+
+type kubeResourceList struct {
+	Items []kubeResource `yaml:"items" json:"items"`
+}
+
+type resourceUnmarshaller func(in []byte, out interface{}) (err error)
+
+type kubernetesResourceOmitter struct {
+	apiVersion   string
+	resourceKind string
+	namespaces   map[string]struct{}
+}
+
+func (k *kubernetesResourceOmitter) File(_, _ string) (bool, error) {
+	return false, nil
+}
+
+func (k *kubernetesResourceOmitter) Contents(path string) (bool, error) {
+	input, err := ioutil.ReadFile(path)
+	if err != nil {
+		return false, err
+	}
+
+	var unmarshaller resourceUnmarshaller
+	switch {
+	case strings.HasSuffix(path, ".yml") || strings.HasSuffix(path, ".yaml"):
+		unmarshaller = yaml.Unmarshal
+	case strings.HasSuffix(path, ".json"):
+		unmarshaller = json.Unmarshal
+	default:
+		return false, nil
+	}
+	var resource kubeResource
+	var resourceList kubeResourceList
+	err = unmarshaller(input, &resource)
+	if err != nil {
+		// this means that the input is not a kubernetes resource
+		return false, nil
+	}
+
+	// check if the input was a list type
+	if strings.HasSuffix(resource.Kind, "List") && resource.ApiVersion == "v1" {
+		err = unmarshaller(input, &resourceList)
+		if err != nil {
+			return false, err
+		}
+	} else {
+		resourceList = kubeResourceList{Items: []kubeResource{resource}}
+	}
+
+	if len(resourceList.Items) == 0 {
+		return false, nil
+	}
+
+	var found bool
+	// loop over the resources and if one of them matches the criteria then set the `found` flag.
+	for _, r := range resourceList.Items {
```

Comment:
whole thing /lgtm, what do you think of pulling this whole k8s stuff above a level higher into the traversal? 
Ideally we should only obfuscate the values of the yaml and not a line-based version of it, at least that was the superpower over sos-cleaner. We can then also build specific obfuscators and omitters based on that schema - think of a configmap filter that looks for certificates in values which is not that easy line-by-line.

Reply:
Omitters read the entire file because they need the whole file to determine if something is valid kubernetes resource. We can have more complex filter of course but that's not omission, it's more obfuscation.

Reply:
exactly, nothing to fix right here. I'm just thinking about pulling the parsing logic a level up, so we can offer a schema/type supported way of `Obfuscator.Contents`. Omission can then leverage the same interface.

basically an addition to the line-by-line processor here: https://github.com/openshift/must-gather-clean/blob/main/pkg/traversal/traversal.go#L108-L128

Reply:
I get it now. So there could be obfuscators which could work on parts of the resource schema this way. It's possible. But I would like to see more user-stories before we commit to something.

Reply:
sounds good, I'll write some tomorrow